### PR TITLE
Bring Parable to 1.0.0

### PR DESCRIPTION
**Pull request by devvoh**
Things to do:

- [x] Make sure all components are under test 100% without relying on any surrounding piece of code, especially Framework.
- [x] Update in-package documentation:
  - [x] Update `README.md` to read like a 1.0.0 release and including examples of current use.
  - [x] Add a `CONTRIBUTING.md` file to explain how to contribute and that not every suggestion will make the cut.
  - [x] Update `CHANGELOG.md` to 1.0.0 final.
- [x] Create Parable Documentation 'site'
  - [x] Build ParaDoc (working title) to simply generate a documentation site, using Parable itself as a base.
  - [x] Write the documentation articles.
    - [x] Installation
    - [x] Introduction
    - [x] Using Parable
      - [x] Overview
      - [x] ...the rest
  - [x] Put it online at https://devvoh.com/parable/docs
- [x] Parable website
  - [x] Make https://devvoh.com/parable redirect to the github repo for now.
  - [x] Set up a simple introductory website with download, documentation and github links.
  - [x] Remove the redirect

Optional: 

- [ ] Check in about that logo (see #24)

### CODE REVIEW

**Comment by devvoh** [comment]
Yo, @Caelaris, @jerry1970, @dmvdbrugge, @RickWong and @FaaPz!

Tagging y'all since I know you guys well enough to directly ask for feedback. So here goes.

I'd add some of you as reviewers, but that includes providing repo Collaborator access and that's a little far for my tastes 😉

Feedback on these changes would be much appreciated! I'll be working on and off the next month on the check list, but if there's anything you feel strongly about so far, let me know 🙂

**Comment by RickWong** [comment]
😎🎉⚡️ nice

**Comment by jerry1970** [comment]
Congrats! :tada:

**Comment by devvoh** [comment]
It's not a release yet, @jerry1970 😉

I created this PR mostly because I'd like to finally do all those things I haven't had to worry about while pre-releasing this thing for years 😂

**Comment by devvoh** [comment]
Looking for:
- Examples of "good" documentation (looking at both tone and technical depth to go into)
- Any suggestions to limit inconsistencies in how the components work.
- Any suggestions for missing functionality (but try to limit it to not extending the scope of Parable by suggesting).

Etcetera. Anything you can think of, this is the moment to let it all out!

Man, this can't go wrong 😅

File: src/Event/Hook.php [hunk 1]
```diff
@@ -8,7 +8,7 @@ class Hook
     protected $hooks = [];
 
     /**
-     * Add hook referencing $closure to $event, returns false if $callable isn't a function
+     * Add hook referencing $closure to $event, returns false if $callable isn't a function.
```
Comment (dmvdbrugge) [inline src/Event/Hook.php]: I think the second half of this comment needs a rewrite, because in no way it returns false.

File: src/Framework/App.php [hunk 2]
```diff
@@ -4,20 +4,21 @@
...
-    const HOOK_ROUTE_MATCH_AFTER    = "parable_route_match_after";
-    const HOOK_SESSION_START_BEFORE = "parable_session_start_before";
-    const HOOK_SESSION_START_AFTER  = "parable_session_start_after";
-    const HOOK_LOAD_ROUTES_BEFORE   = "parable_load_routes_before";
-    const HOOK_LOAD_ROUTES_AFTER    = "parable_load_routes_after";
-    const HOOK_INIT_DATABASE_BEFORE = "parable_init_database_before";
-    const HOOK_INIT_DATABASE_AFTER  = "parable_init_database_after";
+    const PARABLE_VERSION                  = '1.0.0';
+
+    const HOOK_HTTP_404                    = "parable_http_404";
```
Comment (dmvdbrugge) [inline src/Framework/App.php]: Shouldn't these be either alphabetical or in the exact order they're called? (Assuming whatever document generator doesn't do that, if it does then forget it.)

File: src/Event/Hook.php (hunk 1 above)
Comment (devvoh) [reply src/Event/Hook.php]: Agreed. Good find, will go through comments and try to catch all these mismatches 👍

File: src/Framework/App.php (hunk 2 above)
Comment (devvoh) [reply src/Framework/App.php]: Honestly I'm pretty sure there _is_ logic behind this madness, but I can't come up with it. That means I'll make this look prettier ;)

Comment (devvoh) [reply src/Framework/App.php]: I've re-ordered them mostly alphabetically, but I've kept the before/during/after order in there was well.

File: src/Console/Output.php [hunk 3]
```diff
@@ -70,19 +90,23 @@ public function writeln($lines)
...
     {
-        $this->lineLength = 0;
+        $this->disableClearLine();
 
         echo str_repeat(PHP_EOL, $count);
         return $this;
     }
 
     /**
+     * Move the cursor forward by $characters places and reset the lineLength.
```
Comment (dmvdbrugge) [inline src/Console/Output.php]: You can remove the ` and reset the lineLength` here and all the other places, now that lineLength is gone.
Should be replaced with ` and disables line clearing` in the places (not this one) where it does.

**Comment by devvoh** [comment]
Merged in @dmvdbrugge's PR #29 for dynamic return types plugin for phpstorm.

File: dynamicReturnTypeMeta.json [hunk 4]
```diff
@@ -14,12 +14,6 @@
       "class": "\\Parable\\DI\\Container",
       "method": "createAll",
       "position": 0
-    },
-    {
```
Comment (dmvdbrugge) [inline dynamicReturnTypeMeta.json]: Why do you remove this? You have lots of tests using createPartialMock...

Comment (devvoh) [reply dynamicReturnTypeMeta.json]: The better question is, 'why was this removed?'. And this one snuck in because a call to `init-structure` will move the structure's `dynamicReturnTypeMeta.json` not to where it should be  but one folder up.

Which overwrites this one, which is exactly the same except for the `createPartialMock` stuff.

I'm looking into why your call to `copy` it over goes so oddly wrong 👍

Comment (dmvdbrugge) [reply dynamicReturnTypeMeta.json]: > not to where it should be but one folder up.
It goes exactly where it should be, namely the root of the repo. (Which is the correct place, as proven by the override.) You shouldn't init-structure anymore in this repo itself ;)

Comment (devvoh) [reply dynamicReturnTypeMeta.json]: Bug in `Path::getDir()` detected. Will fix so this can't re-occurr 👍

File: src/Console/Command.php [hunk 5]
```diff
@@ -89,80 +125,66 @@ public function getCallable()
...
-    {
-        $this->options[$name] = [
-            'name'          => $name,
-            'required'      => $required,
-            'valueRequired' => $valueRequired,
-            'defaultValue'  => $defaultValue,
-        ];
+    public function addOption(
+        $name,
+        $required = Parameter::PARAMETER_REQUIRED,
```
Comment (dmvdbrugge) [inline src/Console/Command.php]: Default of option should be optional, not required.

File: src/Console/Parameter.php [hunk 6]
```diff
@@ -57,52 +64,44 @@ public function getParameters()
...
-            } elseif ($optionName) {
-                $this->parsedOptions[$optionName] = $parameter;
-                $optionName = null;
             } else {
-                $this->rawArguments[] = $parameter;
+                // For arguments, we need to see if the first one is the command name or not.
+                if ($this->commandNameEnabled && !$this->commandName) {
+                    $this->commandName = $parameter;
+                } else {
+                    $this->arguments[count($this->arguments)] = $parameter;
```
Comment (dmvdbrugge) [inline src/Console/Parameter.php]: Doesn't `$this->arguments[] = $parameter;` suffice?

File: src/Console/Command.php (hunk 5 above)
Comment (devvoh) [reply src/Console/Command.php]: Dangit :P

File: src/Console/Parameter.php (hunk 6 above)
Comment (devvoh) [reply src/Console/Parameter.php]: Vestigial. Orders were 1 and up for a while.

File: tests/Components/Console/ParameterTest.php [hunk 7]
```diff
@@ -218,26 +229,32 @@ public function testArgumentsWorkProperly()
             'command-to-run',
             'argument1',
             'argument2 is a string',
-            '--option1',
-            'value1',
-            'argument3!'
+            '--option1=value1',
+            'argument3!',
+            '--option2=value2',
```
Comment (dmvdbrugge) [inline tests/Components/Console/ParameterTest.php]: Seeing as you made option1 now also the x=y option, shouldn't option 2 now be the "no value" option for full test coverage? (Or add a new `--option3` between arguments 4 and 5.)

Comment (devvoh) [reply tests/Components/Console/ParameterTest.php]: This test is meant to test whether arguments are picked up in order correctly. A value-added option wouldn't add much :smiley:

Comment (dmvdbrugge) [reply tests/Components/Console/ParameterTest.php]: Ok in that case I'm creating more work: where is the test that tests if value-less options work / are parsed correctly?

Comment (devvoh) [reply tests/Components/Console/ParameterTest.php]: `testParseParametersWorkedCorrectly` throws in all possible values:

```
        $this->parameter->setParameters([
            './test.php',
            'command-to-run',
            '--option',
            'argument',
            '--key=value2',
        ]);
```

Comment (dmvdbrugge) [reply tests/Components/Console/ParameterTest.php]: Oh right, sorry... I was thinking about the `PROHIBITED` we had discussed earlier, but we already killed that, my bad. This is good 😄

Comment (devvoh) [reply tests/Components/Console/ParameterTest.php]: A valid question, I had to look it up too :smile:

File: src/Console/Parameter.php [hunk 8]
```diff
@@ -210,82 +174,150 @@ public function checkArguments()
...
+     *
      * @return array
      */
     public function getArguments()
     {
         $returnArray = [];
-        foreach ($this->parsedArguments as $key => $argument) {
-            $returnArray[$key] = $this->getArgument($key);
+        foreach ($this->commandArguments as $argument) {
+            $returnArray[$argument->getName()] = $this->getArgument($argument->getName());
```
Comment (dmvdbrugge) [inline src/Console/Parameter.php]: Shouldn't this just be `$returnArray[$argument->getName()] = $argument->getValue();`?

File: src/Console/Parameter.php [hunk 9]
```diff
@@ -210,82 +174,150 @@ public function checkArguments()
...
+     *
      * @return array
      */
     public function getOptions()
     {
         $returnArray = [];
-        foreach ($this->parsedOptions as $key => $option) {
-            $returnArray[$key] = $this->getOption($key);
+        foreach ($this->commandOptions as $option) {
+            $returnArray[$option->getName()] = $this->getOption($option->getName());
```
Comment (dmvdbrugge) [inline src/Console/Parameter.php]: Same as in `getArguments`

File: src/Console/Parameter.php (hunk 8 above)
Comment (devvoh) [reply src/Console/Parameter.php]: I liked it for consistency, but it can go until `getArgument` decides to do more :)

**Comment by devvoh** [comment]
I'm working on documentation now, so as soon as that's 'done enough' I'll start finishing up all the remaining bits 'n bobs.

I'm aiming for a January 15th release 👍

**Comment by devvoh** [comment]
**NOTE**

The commit titled "Many changes - WILL BREAK EVERY IMPLEMENTATION SO FAR. Check CHANGELOG." does _not lie about that!_

991ad9b will actually break every single implementation out there, and that's by design!

`Bootstrap.php` has been removed! 

Check the updated `index.php_struct` and copy that over your own!

It's very likely that for Console-only apps (I'm looking at you, @dmvdbrugge!), you'll need to do some additional setup in your own executable script (the one that sets up `Console\App` and adds Commands to it), but if you're using the default `parable` script, you should be good, as I fixed that one up already.

Sorry and huzzah, because this is a cleanup I've been wanting for quite some time.

Also, and I can't stress how content I am with this, the following is now possible:

```
<?php
require_once "../vendor/autoload.php";

$app      = \Parable\DI\Container::create(\Parable\Framework\App::class);
$response = \Parable\DI\Container::get(\Parable\Http\Response::class);

$app->get('/hello/{name}', function ($name) use ($response) {
    $response->setContent("Hello, {$name}!");
    $response->send();
});

$app->run();
```

For those _extremely_ light-weight apps that people who seem to like Slim and Silex seem to like to make. Now Parable can do that too. I think I've found at least part of my target audience 😄

**Comment by dmvdbrugge** [comment]
I'm using the parable one, haven't you seen my awesome shell script "shortcut"? 😜 

On other note: very nice those "micro-server-functions" (or how do you call them?)

I do have one remark though: In my opinion, `any` shouldn't need to respecify that you truly want all to match. `get`/`post`/etc are all "shortcuts" to `any`, but I propose they should be shortcuts to `some` which then does take a methods param, at which point `any` could ditch the methods param so it truly matches any, and it too can be a shortcut to `some`.

TLDR: Awesome feature, but rename `any` to `some` and create a new `any`:
```php
    public function any($url, $callable, $templatePath = null)
    {
        $this->some([], $url, $callable, $templatePath);
        return $this;
    }
```

File: src/Framework/App.php [hunk 10]
```diff
@@ -61,55 +70,111 @@ public function __construct(
...
         $this->toolkit    = $toolkit;
         $this->hook       = $hook;
+        $this->path       = $path;
         $this->router     = $router;
         $this->response   = $response;
         $this->url        = $url;
         $this->session    = $session;
         $this->database   = $database;
+
+        $basedir = realpath(__DIR__ . "/../../../../../");
```
Comment (dmvdbrugge) [inline src/Framework/App.php]: Not using `DIRECTORY_SEPARATOR` here?

File: src/Framework/App.php [hunk 11]
```diff
@@ -61,55 +70,111 @@ public function __construct(
...
+        // Enable error reporting if debug is set to true
+        if ($this->config->get('parable.debug') === true) {
+            $this->setErrorReportingEnabled(true);
+        } else {
+            $this->setErrorReportingEnabled(false);
+        }
+
+        // Set the basePath on the url based on the config
+        if ($this->config->get('parable.app.homedir')) {
+            $homedir = trim($this->config->get('parable.app.homedir'), "/");
```
Comment (dmvdbrugge) [inline src/Framework/App.php]: Not using `DIRECTORY_SEPARATOR` here?

Comment (dmvdbrugge) [reply src/Framework/App.php]: Also, there's a lot more places, not gonna point them all out in the PR 😆

**Comment by devvoh** [comment]
And we're back! `DS` and `BASEDIR` are back in because `DIRECTORY_SEPARATOR` is just too damn unwieldy. Additionally, having to define the base directory everywhere, but based on different paths, was not a nice solution.

So now `defines.php` has been added. This script is automatically loaded by composer and will set `DS` and `BASEDIR` appropriately.

File: src/defines.php [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<?php
+// Here we define some constants Parable has a use of
+if (!defined("BASEDIR")) {
+    define("BASEDIR", realpath(__DIR__ . "/../../../.."));
```
Comment (dmvdbrugge) [inline src/defines.php]: Sooo... no `DS` here, just slashes? (Pro-tip: DS is available when you flip around the defines 😉 )

File: structure/public/index.php_struct [hunk 13]
```diff
@@ -1,11 +1,11 @@
 <?php
-$vendorDir = "###VENDOR_PATH###";
+$autoloaderPath = realpath(__DIR__ . "/../vendor/autoload.php");
```
Comment (dmvdbrugge) [inline structure/public/index.php_struct]: `DIRECTORY_SEPARATOR`?

File: src/defines.php (hunk 12 above)
Comment (devvoh) [reply src/defines.php]: Ugh. That's one for tomorrow 😭
